### ISSUE TITLE
Discovery: Improve CLI integration

### ISSUE THREAD

**Issue by swissspidy** [ampproject/amp-wp#1815]
The plugin currently implements a pair of WP-CLI commands for AMP validation: `wp amp validate-site`  and `wp amp reset-site-validation`.

As a WP-CLI maintainer, there are a few things bugging me about this command. For example:

* The `AMP_CLI` class does not extend the `WP_CLI_Command` class as one should.
* Everything is `public static`, just for the tests. Instead, this functionality should be extracted from CLI and/or the CLI command should be tested using WP-CLI Behat tests
* It's not easy to add more CLI commands because everything is in one huge class
* Not following best practices like naming of commands (WP-CLI usually uses `wp <noun> <verb>`), adding debug statements, and others.

**Comment by westonruter** [comment]
> * Not following best practices like naming of commands (WP-CLI usually uses `wp <noun> <verb>`)

Don't `wp amp validate-site` and `wp amp reset-site-validation` follow this practice?

**Comment by swissspidy** [comment]
If you leave the `amp` part out of the picture, it's just two verbs, but they're actually about the same thing — validation. When you compare that to `wp post create` and `wp post delete`, something like `wp amp validation run` and `wp amp validation reset` would fit this more closely. This way you could extend this more easily too, e.g. with something like `wp amp validation info`.

**Comment by swissspidy** [comment]
> The CLI command should be tested using WP-CLI Behat tests

See https://github.com/wearerequired/traduttore/pull/124 for precedence. It's not so easy at the moment so needs testing and perhaps even upstream changes.

**Comment by swissspidy** [comment]
Sort of related: [AMP-Toolbox CLI](https://github.com/ampproject/amp-toolbox/tree/master/packages/cli)

**Comment by schlessera** [comment]
There is duplication of business logic in the WP-CLI commands, like doing database requests and filtering them, for example.

I would suggest creating a repository for dealing with the validation custom post type and the validation custom taxonomy, so that the exact queries and their provenance is abstracted away and deduplicated. This repository would have the following benefits:
- Deduplication of business logic
- Abstraction decouples consuming code from actual implementation
- Easy to change technical details later on
- Easy to swap from CPT to custom DB table if needed
- Outside code can reuse the repository without requiring inside knowledge
- Performance optimizations can be added transparently after the fact

Is this something that makes sense, or are abstractions of that kind not wanted? If this makes sense, I'll create a new issue to discuss the actual interface.

**Comment by schlessera** [comment]
@swissspidy Trying to get Behat tests to work and looking through your prior work on Traduttore right now.

I've got the tests working now, but it ain't pretty. This is the gist of it:
```json
  "scripts": {
    "behat": [
      "rm -rf test-build && mkdir test-build",
      "cp -R {*.php,.distignore,composer.json,assets,back-compat,includes,patches,templates,vendor} test-build",
      "rm -rf test-build/vendor",
      "composer -dtest-build install --no-dev -o",
      "wp dist-archive test-build test-build/amp.zip",
      "behat"
    ]
  }
```
I am currently working on a PR with refactorings, but I will create a separate PR for this Behat configuration, as I assume it will be a longer WIP...

**Comment by schlessera** [comment]
Extract two issues out of this, so once #3056 is merged, this can be closed:

- #3076 - Add functional tests
- #3077 - Build abstraction layer
